include/raft.h

```diff
@@ -47,22 +59,22 @@ enum {
  */
 RAFT_API const char *raft_strerror(int errnum);
 
-typedef unsigned long long raft_id;
+typedef uint64_t raft_id;
 
 /**
- * Hold the value of a raft term. Guaranteed to be at least 64-bit long.
+ * Hold the value of a raft term.
  */
-typedef unsigned long long raft_term;
+typedef uint64_t raft_term;
 
 /**
- * Hold the value of a raft entry index. Guaranteed to be at least 64-bit long.
+ * Hold the value of a raft entry index.
  */
-typedef unsigned long long raft_index;
+typedef uint64_t raft_index;
```

Comment:
I'm not entirely convinced by these type changes. They seem to make a difference only if the underlying value exceeds the 64bit threshold, however in that case we just shift the problem because the value will overflow and restart from 0, regardless of the fact that it can be serialized "correctly".

I feel the proper fix would be to check that we don't exceed the 64bit threshold when those values are generated by the lib (in case of raft_term, raft_index, raft_time) or provided by the user (in case of raft_time).

Also, these are somehow abstract types and serialization concerns should be kept separate.

Reply:
Also, in principle a certain architecture could have a "native" hardware type for ```unsigned long long``` that is more efficient than ```uint64_t```. This is of course purely speculative and it most probably wouldn't have a practical impact of any kind (performance or otherwise), but it's just to it's just to illustrate that generic C types are usually a better choice because the compiler has more freedom to map it to the most appropriate hardware type that does the job.

Reply:
I'll add that in practical terms, `raft_time` will take >500 million years to overflow, same time frame for `raft_index` if we consider a steady flow of one transaction per millisecond. For `raft_term` overflowing the headroom is even higher, and for `raft_id` the `raft_digest()` should be generally used (or some other user-defined method to avoid duplicates) and there too the headroom is even higher as well because configuration changes are even more sparse than term bumps.

All in all, I'd suggest to keep things as they are.

Reply:
The thing I'd like to make sure here is that we don't use types that may change size depending on architecture as if that makes it over the wire or to disk, we may hit some issues.

Reply:
I think for the `raft_time`, `raft_term` and `raft_index` I would leave things as-is then, not even checking overflow as it's so improbable. For the `raft_id` I think it still makes sense to change the type to `uint64_t`, we (hopefully) get free compiler warnings that way if, on some exotic system, the user would ever assign a larger value to a `raft_id`. And we don't have to check for user supplied values that overflow that way.

Reply:
> The thing I'd like to make sure here is that we don't use types that may change size depending on architecture as if that makes it over the wire or to disk, we may hit some issues.

Serialization is indeed a separate concern, as I was mentioning. The `unsigned long long` type is guaranteed to be at least 64bit on any architecture (and the compiler can use the hardware type that is best fit for that). We know that practically speaking we won't overflow those 64bits (and even if that was a possible thing we should add boundary checks instead, regardless). When we serialize those values we always use an architecture-independent encoding with specific sizes.

Reply:
> I think for the `raft_time`, `raft_term` and `raft_index` I would leave things as-is then, not even checking overflow as it's so improbable. For the `raft_id` I think it still makes sense to change the type to `uint64_t`, we (hopefully) get free compiler warnings if, on some exotic system, the user would ever assign a larger value to a `raft_id`. And we don't have to check for user supplied values that overflow.

I'm not sure to understand this. This a runtime concern no? If a user assigns a larger value, it would just overflow and wrap around at run time, with no compiler warnings whatsoever.

If we find that a concern (which at least to me doesn't seem to be the case, practically speaking), I'd suggest to introduce a boundary check. I believe adding that check in `raft_bootstrap()` and `raft_add()` would be enough, since if I'm not mistaking those are the only ways to introduce new IDs.

Reply:
Yeah looks like I was overestimating the C typesystem and that something like:

```C
#include <inttypes.h>

typedef uint8_t raft_id;

static void f(raft_id id)
{
    (void) id;
}

int main(int argc, char **argv)
{
   uint64_t id_too_large = (uint64_t)1 << 63;
   f(id_too_large);
   return 0;
}
```

would let the compiler complain, but it doesn't.

Reply:
> Yeah looks like I was overestimating the C typesystem and that something like:
> 
> ```c
> #include <inttypes.h>
> 
> typedef uint8_t raft_id;
> 
> static void f(raft_id id)
> {
>     (void) id;
> }
> 
> int main(int argc, char **argv)
> {
>    uint64_t id_too_large = (uint64_t)1 << 63;
>    f(id_too_large);
>    return 0;
> }
> ```
> 
> would let the compiler complain, but it doesn't.

Yeah. If you want to introduce a boundary check, actually the best place would not be what I mentioned above but rather the `configurationEncodeToBuf()` function in `configuration.c` where we actually convert the `unsigned long long` to a 64bit little-endian representation. Adding the check there makes sense since this is really a serialization concern from our perspective. In terms of what value the user chooses, we have no control on that, if it overflows `unsigned long long` that's a user concern, not ours. Plus, the recommended way to generate an ID is with `raft_digest()`, which is safe.